File: backend/app/shared/domain/money.py
"""Money value object — the one real invariant in this domain.

You must never add 100 EUR to 100 USD as if they were the same number. `Money`
enforces that: same-currency arithmetic is allowed; cross-currency arithmetic
raises. Converting between currencies is *explicit* — you pass a rate and the
target currency — so a conversion can never happen by accident.

Amounts are stored rounded to the cent (via the shared `round2`) so a Money and
its equivalent literal compare equal.
"""

from __future__ import annotations

from dataclasses import dataclass

from app.shared.domain.rounding import round2


class CurrencyMismatchError(ValueError):
    """Raised when two different currencies are combined without conversion."""

    def __init__(self, a: str, b: str) -> None:
        super().__init__(
            f"Refusing implicit cross-currency arithmetic: {a} vs {b}. "
            "Convert to a single currency first."
        )


def _validate_currency(code: str) -> str:
    if not isinstance(code, str) or len(code) != 3 or not code.isalpha():
        raise ValueError(f"Invalid ISO 4217 currency code: {code!r}")
    return code.upper()


@dataclass(frozen=True)
class Money:
    amount: float
    currency: str

    def __post_init__(self) -> None:
        # frozen dataclass: mutate through object.__setattr__ during init only.
        object.__setattr__(self, "currency", _validate_currency(self.currency))
        object.__setattr__(self, "amount", round2(float(self.amount)))

    def _same_currency(self, other: Money) -> None:
        if self.currency != other.currency:
            raise CurrencyMismatchError(self.currency, other.currency)

    def add(self, other: Money) -> Money:
        self._same_currency(other)
        return Money(self.amount + other.amount, self.currency)

    def subtract(self, other: Money) -> Money:
        self._same_currency(other)
        return Money(self.amount - other.amount, self.currency)

    def scale(self, factor: float) -> Money:
        """Multiply by a scalar (e.g. quantity) — stays in the same currency."""
        return Money(self.amount * factor, self.currency)

    def convert(self, rate: float, to_currency: str) -> Money:
        """Explicitly convert to another currency at a given rate."""
        return Money(self.amount * rate, to_currency)

    def is_zero(self) -> bool:
        return self.amount == 0.0

    def __lt__(self, other: Money) -> bool:
        self._same_currency(other)
        return self.amount < other.amount

    def __le__(self, other: Money) -> bool:
        self._same_currency(other)
        return self.amount <= other.amount

    def __gt__(self, other: Money) -> bool:
        self._same_currency(other)
        return self.amount > other.amount

    def __ge__(self, other: Money) -> bool:
        self._same_currency(other)
        return self.amount >= other.amount

    def __str__(self) -> str:
        return f"{self.amount:.2f} {self.currency}"


def zero(currency: str) -> Money:
    return Money(0.0, currency)

File: frontend/src/shared/domain/money.ts
/**
 * Money value object — the one real invariant in this domain.
 *
 * You must never add 100 EUR to 100 USD as if they were the same number. Money
 * enforces that: same-currency arithmetic is allowed; cross-currency arithmetic
 * throws. Conversion is explicit (pass a rate + target currency) so it can
 * never happen by accident. Mirrors backend/app/shared/domain/money.py.
 */

import { round2 } from "./rounding";

export class CurrencyMismatchError extends Error {
  constructor(a: string, b: string) {
    super(
      `Refusing implicit cross-currency arithmetic: ${a} vs ${b}. ` +
        "Convert to a single currency first.",
    );
    this.name = "CurrencyMismatchError";
  }
}

function validateCurrency(code: string): string {
  if (typeof code !== "string" || !/^[A-Za-z]{3}$/.test(code)) {
    throw new Error(`Invalid ISO 4217 currency code: ${JSON.stringify(code)}`);
  }
  return code.toUpperCase();
}

export class Money {
  readonly amount: number;
  readonly currency: string;

  constructor(amount: number, currency: string) {
    this.currency = validateCurrency(currency);
    this.amount = round2(Number(amount));
    Object.freeze(this);
  }

  private sameCurrency(other: Money): void {
    if (this.currency !== other.currency) {
      throw new CurrencyMismatchError(this.currency, other.currency);
    }
  }

  add(other: Money): Money {
    this.sameCurrency(other);
    return new Money(this.amount + other.amount, this.currency);
  }

  subtract(other: Money): Money {
    this.sameCurrency(other);
    return new Money(this.amount - other.amount, this.currency);
  }

  /** Multiply by a scalar (e.g. quantity) — stays in the same currency. */
  scale(factor: number): Money {
    return new Money(this.amount * factor, this.currency);
  }

  /** Explicitly convert to another currency at a given rate. */
  convert(rate: number, toCurrency: string): Money {
    return new Money(this.amount * rate, toCurrency);
  }

  equals(other: Money): boolean {
    return this.currency === other.currency && this.amount === other.amount;
  }

  isZero(): boolean {
    return this.amount === 0;
  }

  lessThan(other: Money): boolean {
    this.sameCurrency(other);
    return this.amount < other.amount;
  }

  greaterThan(other: Money): boolean {
    this.sameCurrency(other);
    return this.amount > other.amount;
  }

  toString(): string {
    return `${this.amount.toFixed(2)} ${this.currency}`;
  }
}

export function zeroMoney(currency: string): Money {
  return new Money(0, currency);
}
